File: frontend/src/app/core/services/loan-officer.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  OfficerDashboardResponse,
  LoanApplicationSummaryOfficer,
  RecentActivityOfficer,
  LoanApplicationResponse,
  CompleteApplicationDetailsResponse,
  ApplicationInfo,
  ApplicantIdentity,
  PersonalDetailsNested,
  AddressInfo,
  ContactInfo,
  VerificationStatusType,
  EmploymentDetails,
  CompanyContact,
  BankDetails,
  EmploymentVerificationStatus,
  FinancialAssessment,
  LoanDetails,
  ExistingLoan,
  CalculatedRatios,
  RiskAssessment,
  VerificationSummary,
  PersonalDetails,
  FinancialDetails,
  DocumentInfo,
  ExternalVerificationInfo,
  AuditEntry,
  DocumentVerificationRequest,
  DocumentResubmissionRequest,
  DocumentResubmissionResponse,
  LoanDecisionRequest,
  LoanDecisionResponse,
  ComplianceFlagRequest,
  ExternalVerificationResponse
} from '../models/officer.model';


@Injectable({
  providedIn: 'root'
})
export class LoanOfficerService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/officer`;

  /**
   * Get loan officer dashboard with statistics
   */
  getDashboard(): Observable<OfficerDashboardResponse> {
    return this.http.get<OfficerDashboardResponse>(`${this.apiUrl}/dashboard`);
  }

  /**
   * Get applications assigned to current officer
   */
  getAssignedApplications(): Observable<LoanApplicationResponse[]> {
    return this.http.get<LoanApplicationResponse[]>(`${this.apiUrl}/assigned-applications`);
  }

  /**
   * Get specific application for review
   */
  getApplicationForReview(applicationId: string): Observable<LoanApplicationResponse> {
    return this.http.get<LoanApplicationResponse>(`${this.apiUrl}/applications/${applicationId}`);
  }

  /**
   * Get complete application details with all sections
   */
  getCompleteApplicationDetails(applicationId: string): Observable<CompleteApplicationDetailsResponse> {
    return this.http.get<CompleteApplicationDetailsResponse>(
      `${this.apiUrl}/applications/${applicationId}/complete-details`
    );
  }

  /**
   * Start document verification process
   */
  startDocumentVerification(applicationId: string): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(
      `${this.apiUrl}/applications/${applicationId}/start-verification`,
      {}
    );
  }

  /**
   * Complete document verification
   */
  verifyDocuments(
    applicationId: string,
    request: DocumentVerificationRequest
  ): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(
      `${this.apiUrl}/applications/${applicationId}/verify-documents`,
      request
    );
  }

  /**
   * Trigger external verification (fraud detection & credit scoring)
   */
  triggerExternalVerification(applicationId: string): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(
      `${this.apiUrl}/applications/${applicationId}/trigger-external-verification`,
      {}
    );
  }

  /**
   * Complete external verification and get results
   */
  completeExternalVerification(applicationId: string): Observable<ExternalVerificationResponse> {
    return this.http.post<ExternalVerificationResponse>(
      `${this.apiUrl}/applications/${applicationId}/complete-external-verification`,
      {}
    );
  }

  /**
   * Get applications ready for final decision (EXCLUDES compliance applications)
   */
  getApplicationsReadyForDecision(): Observable<LoanApplicationResponse[]> {
    return this.http.get<LoanApplicationResponse[]>(`${this.apiUrl}/ready-for-decision`);
  }

  /**
   * Get all post-compliance applications
   * Includes: FLAGGED_FOR_COMPLIANCE, UNDER_INVESTIGATION, AWAITING_COMPLIANCE_DECISION, READY_FOR_DECISION (from compliance)
   */
  getPostComplianceApplications(): Observable<LoanApplicationResponse[]> {
    return this.http.get<LoanApplicationResponse[]>(`${this.apiUrl}/applications/post-compliance`);
  }

  /**
   * Get compliance investigation data for an application
   * Returns the detailed JSON investigation data from compliance team
   */
  getComplianceInvestigationData(applicationId: string): Observable<any> {
    return this.http.get(`${this.apiUrl}/applications/${applicationId}/compliance-investigation`);
  }

  /**
   * Request document resubmission from applicant
   */
  requestDocumentResubmission(
    applicationId: string,
    request: DocumentResubmissionRequest
  ): Observable<DocumentResubmissionResponse> {
    return this.http.post<DocumentResubmissionResponse>(
      `${this.apiUrl}/applications/${applicationId}/request-resubmission`,
      request
    );
  }

  /**
   * Get audit trail for an application
   */
  getAuditTrail(applicationId: string): Observable<AuditEntry[]> {
    return this.http.get<AuditEntry[]>(
      `${this.apiUrl}/applications/${applicationId}/audit-trail`
    );
  }

  /**
   * Approve loan application
   */
  approveApplication(
    applicationId: string,
    request: LoanDecisionRequest
  ): Observable<LoanDecisionResponse> {
    return this.http.post<LoanDecisionResponse>(
      `${this.apiUrl}/applications/${applicationId}/approve`,
      request
    );
  }

  /**
   * Reject loan application
   */
  rejectApplication(
    applicationId: string,
    request: LoanDecisionRequest
  ): Observable<LoanDecisionResponse> {
    return this.http.post<LoanDecisionResponse>(
      `${this.apiUrl}/applications/${applicationId}/reject`,
      request
    );
  }

  /**
   * Flag application for compliance review
   */
  flagForCompliance(
    applicationId: string,
    request: ComplianceFlagRequest
  ): Observable<LoanDecisionResponse> {
    return this.http.post<LoanDecisionResponse>(
      `${this.apiUrl}/applications/${applicationId}/flag-for-compliance`,
      request
    );
  }

  /**
   * Helper method to format currency
   */
  formatCurrency(amount: number): string {
    return new Intl.NumberFormat('en-IN', {
      style: 'currency',
      currency: 'INR',
      maximumFractionDigits: 0
    }).format(amount);
  }

  /**
   * Helper method to format date
   */
  formatDate(date: Date | string): string {
    return new Date(date).toLocaleDateString('en-IN', {
      year: 'numeric',
      month: 'short',
      day: 'numeric'
    });
  }

  /**
   * Get display status for loan officer - freezes compliance statuses
   * Loan officer sees FLAGGED_FOR_COMPLIANCE for all compliance-related statuses
   */
  getDisplayStatus(actualStatus: string): string {
    if (!actualStatus) return '';

    // List of all compliance-related statuses from backend ApplicationStatus enum
    const complianceStatuses = [
      'FLAGGED_FOR_COMPLIANCE',
      'COMPLIANCE_REVIEW',
      'PENDING_COMPLIANCE_DOCS',
      'COMPLIANCE_TIMEOUT',
      'UNDER_INVESTIGATION',
      'AWAITING_COMPLIANCE_DECISION'
    ];

    // If status is any compliance-related status, show as FLAGGED_FOR_COMPLIANCE
    // This "freezes" the status for loan officer during entire compliance process
    if (complianceStatuses.includes(actualStatus)) {
      return 'FLAGGED_FOR_COMPLIANCE';
    }

    return actualStatus;
  }

  /**
   * Helper method to get status badge color
   */
  getStatusBadgeClass(status: string): string {
    const statusColors: { [key: string]: string } = {
      'SUBMITTED': 'bg-blue-100 text-blue-800',
      'UNDER_REVIEW': 'bg-yellow-100 text-yellow-800',
      'DOCUMENT_VERIFICATION': 'bg-purple-100 text-purple-800',
      'PENDING_EXTERNAL_VERIFICATION': 'bg-orange-100 text-orange-800',
      'READY_FOR_DECISION': 'bg-green-100 text-green-800',
      'APPROVED': 'bg-green-500 text-white',
      'REJECTED': 'bg-red-500 text-white',
      'FLAGGED_FOR_COMPLIANCE': 'bg-red-100 text-red-800'
    };
    return statusColors[status] || 'bg-gray-100 text-gray-800';
  }

  /**
   * Helper method to get priority badge color
   */
  getPriorityBadgeClass(priority: string): string {
    const priorityColors: { [key: string]: string } = {
      'HIGH': 'bg-red-100 text-red-800',
      'MEDIUM': 'bg-yellow-100 text-yellow-800',
      'LOW': 'bg-blue-100 text-blue-800'
    };
    return priorityColors[priority] || 'bg-gray-100 text-gray-800';
  }

  /**
   * Helper method to get risk level badge color
   */
  getRiskLevelBadgeClass(riskLevel: string): string {
    const riskColors: { [key: string]: string } = {
      'LOW': 'bg-green-100 text-green-800',
      'MEDIUM': 'bg-yellow-100 text-yellow-800',
      'HIGH': 'bg-red-100 text-red-800'
    };
    return riskColors[riskLevel] || 'bg-gray-100 text-gray-800';
  }

  /**
   * Format date and time for display
   */
  formatDateTime(date: Date | string): string {
    return new Date(date).toLocaleString('en-IN', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    });
  }
}

File: server/src/main/java/com/tss/loan/service/LoanOfficerService.java
package com.tss.loan.service;

import java.util.List;
import java.util.UUID;

import com.tss.loan.dto.request.DocumentResubmissionRequest;
import com.tss.loan.dto.request.DocumentVerificationRequest;
import com.tss.loan.dto.response.AuditLogResponse;
import com.tss.loan.dto.response.CompleteApplicationDetailsResponse;
import com.tss.loan.dto.response.DocumentResubmissionResponse;
import com.tss.loan.dto.response.ExternalVerificationResponse;
import com.tss.loan.dto.response.LoanApplicationResponse;
import com.tss.loan.dto.response.OfficerDashboardResponse;
import com.tss.loan.entity.user.User;

/**
 * Service for loan officer operations
 */
public interface LoanOfficerService {
    
    /**
     * Get dashboard statistics for loan officer
     * @param officer The loan officer
     * @return Dashboard response with statistics
     */
    OfficerDashboardResponse getDashboard(User officer);
    
    /**
     * Get applications assigned to the officer
     * @param officer The loan officer
     * @return List of assigned applications
     */
    List<LoanApplicationResponse> getAssignedApplications(User officer);
    
    /**
     * Get specific application for review (with security check)
     * @param applicationId The application ID
     * @param officer The requesting officer
     * @return Application details
     */
    LoanApplicationResponse getApplicationForReview(UUID applicationId, User officer);
    
    /**
     * Get complete application details with all sections for manual verification
     * @param applicationId The application ID
     * @param officer The requesting officer
     * @return Complete application details with all verification sections
     */
    CompleteApplicationDetailsResponse getCompleteApplicationDetails(UUID applicationId, User officer);
    
    /**
     * Get complete application details without officer validation (for internal service use)
     * @param applicationId The application ID
     * @return Complete application details with all verification sections
     */
    CompleteApplicationDetailsResponse getCompleteApplicationDetailsInternal(UUID applicationId);
    
    /**
     * Start document verification process
     * @param applicationId The application ID
     * @param officer The loan officer
     */
    void startDocumentVerification(UUID applicationId, User officer);
    
    /**
     * Complete document verification
     * @param applicationId The application ID
     * @param request Verification details
     * @param officer The loan officer
     */
    void completeDocumentVerification(UUID applicationId, DocumentVerificationRequest request, User officer);
    
    /**
     * Trigger external verification (fraud detection) after manual verification
     * @param applicationId The application ID
     * @param officer The loan officer
     */
    void triggerExternalVerification(UUID applicationId, User officer);
    
    /**
     * Complete external verification (fraud check) with credit scoring and move directly to READY_FOR_DECISION
     * @param applicationId The application ID
     * @param officer The loan officer
     * @return Enhanced response with credit scoring details
     */
    ExternalVerificationResponse completeExternalVerification(UUID applicationId, User officer);
    
    /**
     * Get applications ready for final decision
     * @param officer The loan officer
     * @return List of applications ready for decision
     */
    List<LoanApplicationResponse> getApplicationsReadyForDecision(User officer);
    
    /**
     * Request document resubmission from applicant with automatic same-officer reassignment
     * @param applicationId The application ID
     * @param request Resubmission details
     * @param officer The loan officer
     * @return Resubmission response with details
     */
    DocumentResubmissionResponse requestDocumentResubmission(UUID applicationId, DocumentResubmissionRequest request, User officer);
    
    /**
     * Get combined audit trail for an application (audit logs + workflow history)
     * @param applicationId The application ID
     * @param officer The loan officer
     */
    List<AuditLogResponse> getApplicationAuditTrail(UUID applicationId, User officer);
    
    /**
     * Get compliance review summary for loan officer
     * Returns complete compliance investigation data, notes, and timeline
     * @param applicationId The application ID
     * @param officer The loan officer
     * @return Compliance review summary
     */
    com.tss.loan.dto.response.ComplianceReviewSummaryResponse getComplianceReviewSummary(UUID applicationId, User officer);
    
    /**
     * Acknowledge compliance review by loan officer
     * Tracks that loan officer has reviewed compliance findings before making decision
     * @param applicationId The application ID
     * @param request Acknowledgment request with optional notes
     * @param officer The loan officer
     */
    void acknowledgeComplianceReview(UUID applicationId, com.tss.loan.dto.request.AcknowledgeComplianceReviewRequest request, User officer);
    
    /**
     * Get all post-compliance applications for loan officer
     * Returns applications that went through compliance process (all statuses)
     * @param officer The loan officer
     * @return List of post-compliance applications
     */
    List<LoanApplicationResponse> getPostComplianceApplications(User officer);
    
    /**
     * Get compliance investigation data for an application
     * Returns the detailed investigation data stored by compliance officer
     */
    String getComplianceInvestigationData(UUID applicationId, User officer);
}
